ecmascript/transforms/compat/src/bugfixes/template_literal_caching.rs

```diff
@@ -0,0 +1,308 @@
+// can cause tricky bugs.
+//   Object``===Object``  // true, should be false.
+//
+// Benchmarks: https://jsperf.com/compiled-tagged-template-performance
+pub fn template_literal_caching() -> impl Fold {
+    TemplateLiteralCaching::default()
+}
+#[derive(Default, Clone)]
+struct TemplateLiteralCaching {
+    decls: Vec<VarDeclarator>,
+    helper_ident: Option<Ident>,
+}
+
+impl TemplateLiteralCaching {
+    fn create_binding(&mut self, name: Ident, init: Option<Expr>) {
+        let init = if let Some(init) = init {
+            Some(Box::new(init))
+        } else {
+            None
+        };
+        self.decls.push(VarDeclarator {
+            span: DUMMY_SP,
+            name: Pat::Ident(BindingIdent::from(name.clone())),
+            init,
+            definite: false,
+        })
+    }
+
+    fn create_var_decl(&mut self) -> Option<Stmt> {
+        if self.decls.len() > 0 {
+            return Some(Stmt::Decl(Decl::Var(VarDecl {
+                span: DUMMY_SP,
+                kind: VarDeclKind::Let,
+                declare: false,
+                decls: self.decls.clone(),
+            })));
+        }
+        None
+    }
+}
+
+impl Fold for TemplateLiteralCaching {
+    noop_fold_type!();
+
+    fn fold_expr(&mut self, n: Expr) -> Expr {
+        let n = n.fold_children_with(self);
+        match n {
+            Expr::TaggedTpl(n) => {
+                if self.helper_ident.is_none() {
+                    // Create an identity function helper:
+                    //   identity = t => t
+                    let helper_ident = private_ident!("_");
+                    let t = private_ident!("t");
+                    self.helper_ident = Some(helper_ident.clone());
+                    self.create_binding(
+                        helper_ident.clone(),
+                        Some(Expr::Arrow(ArrowExpr {
+                            span: DUMMY_SP,
+                            params: vec![Pat::Ident(BindingIdent::from(t.clone()))],
+                            body: BlockStmtOrExpr::Expr(Box::new(Expr::Ident(t.clone()))),
+                            is_async: false,
+                            is_generator: false,
+                            type_params: None,
+                            return_type: None,
+                        })),
+                    )
+                }
+
+                let helper_ident = self.helper_ident.as_ref().unwrap();
+
+                // Use the identity function helper to get a reference to the template's
+                // Strings. We replace all expressions with `0` ensure Strings has
+                // the same shape.   identity`a${0}`
+                let template = TaggedTpl {
+                    span: DUMMY_SP,
+                    tag: Box::new(Expr::Ident(helper_ident.clone())),
+                    quasis: n.quasis,
+                    exprs: n
+                        .exprs
+                        .clone()
```

Comment:
It seems like the value is ignored on next call to `.map`.
If then, we don't need to clone the vector I guess.